core/shared/src/main/scala/spire/math/interval/Overlap.scala

```diff
@@ -0,0 +1,90 @@
+package spire.math.interval
+
+import spire.algebra.{Eq, Order}
+import spire.math.Interval
+
+/**
+  * An ADT that represents overlapping result for any two intervals.
+  * All subclasses and method names are "left-biased":
+  * for example, [[StrictlyLess]] result means, that [[lhs]] is strictly less than [[rhs]].
+  */
+sealed abstract class Overlap[A: Order] extends Serializable {
+  def lhs: Interval[A]
```

Comment:
I think `lhs` and `rhs` are bit oddly named, given that we may re-order the arguments from the original `overlap` call. Personal preference is that we actually get rid of `lhs` and `rhs` completely and expect folks to pattern match. If someone just wants to "order" them, then if we define the `PartialOrder[Interval[A]]` using the `overlap` method, then we can just order them with that instead.